design/ceph/object/ceph-rgw-support-multisite-for-external-cluster.md

```diff
@@ -0,0 +1,50 @@
+---
+# Improvements to RGW Multisite configuration
+target-version: release-1.10
+---
+
+# Feature Name
+ Improvements to RGW Multisite configuration
+
+## Summary
+Following are the major changes planned with this feature:
+* For external RGW servers, have place holder for zone/zonegroup/realm settings in `GatewaySpec`.
+* Provide option to add Endpoints to `zone` configuration.
+* Encode the `Endpoint` information into `realm-system-user` secret using the RGW Kubernetes service name.
+
+### Goals
+Improve usability of RGW mulitisite for customers and make it more friendly for disaster recovery scenarios. And planning to cover following use-cases:
+* two ceph object store clusters running in two different kubernetes cluster.
+* one ceph object store resides in kubernetes and another is external.
```

Comment:
Whether or not the object stores are internal or external, why does it matter? Seems like RGW multisite should behave the same in either case. Although I'm not sure it makes sense for the admin of a rook cluster to configure RGW multisite on an external cluster that is managed by cephadm. The cluster administrator should be the one configuring multisite configs.

Reply:
Agree with @travisn on the first comment. My understanding of external mode is that it's only meant to expose RGWs to things running inside of K8s. When Rook sets this up there is no ceph cluster (Mons and OSDs) it is setting up. Unless my understanding of external mode is inaccurate, whether or not the External RGWs have multisite configured and are syncing data between another cluster doesn't change the way the resource is consumed or viewed by an application.

Reply:
@thotz Here are my comments on your following scenarios:
1. Does OBC and CephObjectStoreUser fail when the Multisite CRDs are created? If so do you know why? That is an issue worth fixing, but I don't see it as a multisite improvement just maybe some added config to those CRDs
2. In this scenario all you would have to do is add an endpoint of an RGW from the master zone in the external Ceph cluster to the `endpoint` in the `pull` section of the `CephObjectRealm` config, and then create the realm and the zone group CRs with the same name as the external Ceph cluster. That's it, as long as the configured endpoint can be reached from the k8s environment.
3. This scenario is practically the same as the 2nd one. The only thing that matters between syncing two clusters, regardless of whether the cluster is configured in K8s or not is that the RGWs are in the realm reachable from one another, and matching the names.

Reply:
> @thotz Here are my comments on your following scenarios:
> 
>     1. Does OBC and CephObjectStoreUser fail when the Multisite CRDs are created? If so do you know why? That is an issue worth fixing, but I don't see it as a multisite improvement just maybe some added config to those CRDs
> 

The simple or common use cases having multiple zones with belongs to a zonegroup work fine for obc/cephobjectstoreuser if the ceph cluster is internal.  
for RGW belongs to external ceph cluster this was falling see upstream issue https://github.com/rook/rook/issues/6342 and ODF bug https://bugzilla.redhat.com/show_bug.cgi?id=2050224. For the ODF he mentioned to add the `internal-rgw-service-name` to zone list and it worked for him. I am not quite sure about that

For situations with zone-groups(use cases with two distant ceph-cluster/k8s cluster), this will be a bit tricky, especially for OBC

>     2. In this scenario all you would have to do is add an endpoint of an RGW from the master zone in the external Ceph cluster to the `endpoint` in the `pull` section of the `CephObjectRealm` config, and then create the realm and the zone group CRs with the same names from the external Ceph cluster. That's it, as long as the configured endpoint can be reached from the k8s environment.
> 
>     3. This scenario is practically the same as the 2nd one. The only thing that matters between syncing two clusters, regardless of whether the cluster is configured in K8s or not is that the RGWs are in the realm reachable from one another, and match the names.

Yeah, I completely agree that the above use-case works fine, but my idea is to improve this, not reimplement the wheel like adding more details to realm secrets

Reply:
@leseb you posted #6342. Do you remember more context about that that could assist us with this design? My understanding of that issue was that we should be able to create realm/zonegroup/zone resources to assist that. Or did you intend some other strategy?

Reply:
AFAIR, https://github.com/rook/rook/issues/6342's goal was to adapt `NewMultisiteContext()` so that `getMultisiteForObjectStore()` reports the correct realm/zonegroup/zone even if the cluster is external. This code:

https://github.com/rook/rook/blob/1acc0621af652b37b10c224064ba8c80df5e02fe/pkg/operator/ceph/object/objectstore.go#L198-L204

needs to change.